homeassistant/components/knx.py

```diff
@@ -0,0 +1,310 @@
+"""
+Support for KNX components.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/knx/
+
+(c) 2016 Daniel Matuschek <info@open-homeautomation.com>
+"""
+import logging
+
+from homeassistant.const import EVENT_HOMEASSISTANT_STOP
+from homeassistant.helpers.entity import Entity
+
+DOMAIN = "knx"
+REQUIREMENTS = ['knxip==0.2.3']
+
+EVENT_KNX_FRAME_RECEIVED = "knx_frame_received"
+
+CONF_HOST = "host"
+CONF_PORT = "port"
+
+DEFAULT_PORT = "3671"
+
+KNXTUNNEL = None
+HASS = None
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def setup(hass, config):
+    """Setup the connection to the KNX IP interface."""
+    global KNXTUNNEL, HASS
+
+    from knxip.ip import KNXIPTunnel
+    from knxip.core import KNXException
+
+    host = config[DOMAIN].get(CONF_HOST, None)
+
+    if host is None:
+        _LOGGER.exception("KNX IP interface hostname undefined")
+        return False
+
+    try:
+        port = int(config[DOMAIN].get(CONF_PORT, DEFAULT_PORT))
+    except ValueError:
+        _LOGGER.exception("Can't parse KNX IP interface port")
+        return False
+
+    if KNXTUNNEL:
+        _LOGGER.warning("Tunnel already exists, not opening another tunnel")
+        return True
+
+    HASS = hass
+    KNXTUNNEL = KNXIPTunnel(host, port)
+    try:
+        KNXTUNNEL.connect()
+    except KNXException as ex:
+        _LOGGER.exception("Can't connect to KNX/IP interface: %s", ex)
+        KNXTUNNEL = None
+        return False
+
+    KNXTUNNEL.notify = knx_message_received
+    _LOGGER.info("KNX IP tunnel to %s:%i established", host, port)
+
+    hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, close_tunnel)
+    return True
+
+
+def knx_message_received(address, data):
```

Comment:
You should define this method inside `setup(hass, config)`. That way you do not have to keep a global reference to the `hass` object.

On top of that, I think that the frames being distributed to the various devices should be handled inside the component instead of passing it over the event bus. So have a class that wraps KNXIPTunnel and allow KNX devices to register with that.